### PR TITLE
Move timestamps to header

### PR DESCRIPTION
**Pull request by CodeSandwich**
Fixes https://github.com/radicle-dev/radicle-registry/issues/475.

~It puts the timestamp in the seal of the block: it's the last piece of digest added to a block. This way the timestamp can be fully controlled by the PoW algorithm: it sets it, validates it and then uses it. Theoretically it could even alter it while mining (Bitcoin style).~

### CODE REVIEW

**Comment by geigerzaehler** [comment]
I think it would be preferable to put the timestamp into a separate block header digest item so that it can be inspected by the runtime when a block executes.

**Comment by CodeSandwich** [comment]
The timestamp is now a separate digest item in the header.

This is a hard-fork-worth breaking change, because now the PoW expects the timestamp to be present in the header. This is ok, because now we're gathering hard changes in master.

File: runtime/src/timestamp_in_digest.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::Moment;
+use frame_system::{DigestItemOf, Trait};
+use parity_scale_codec::{Decode, DecodeAll, Encode, Error};
+use sp_runtime::{ConsensusEngineId, Digest, DigestItem};
+
+const CONSENSUS_ID: ConsensusEngineId = *b"time";
+
+pub fn load<Hash: Decode + Encode>(digest: &Digest<Hash>) -> Option<Result<Moment, Error>> {
```
Comment (geigerzaehler) [inline runtime/src/timestamp_in_digest.rs:23]: This is not used by the on-chain runtime. Could we put it behind the `std` feature?

File: runtime/src/timestamp_in_digest.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    digest
+        .log(|item| match item {
+            DigestItem::Consensus(CONSENSUS_ID, encoded) => Some(encoded),
+            _ => None,
+        })
+        .map(|encoded| DecodeAll::decode_all(encoded))
+}
+
+pub fn digest_item<T: Trait>(timestamp: Moment) -> DigestItemOf<T> {
```
Comment (geigerzaehler) [inline runtime/src/timestamp_in_digest.rs:32]: Could we expand the return type and drop the generic `T`?
```suggestion
pub fn digest_item(timestamp: Moment) -> sp_runtime::DigestItem<crate::Hash> {
```

**Comment by CodeSandwich** [comment]
> I would still prefer to have the timestamp put in the digest explicitly by the author to avoid coupling of runtime and consensus.

This isn't possible without forking the PoW consensus crate, there's zero possible configuration. But if we had full control, would we do it differently? it makes sense to check the timestamp and put it into the digest at the beginning of the inherent flow.

File: runtime/src/timestamp_in_digest.rs (hunk 2 above)
Comment (CodeSandwich) [reply runtime/src/timestamp_in_digest.rs]: Fixed

File: runtime/src/timestamp_in_digest.rs (hunk 1 above)
Comment (CodeSandwich) [reply runtime/src/timestamp_in_digest.rs]: Fixed

**Comment by geigerzaehler** [comment]
>  But if we had full control, would we do it differently?

I imagined that we put the timestamp into the digest when the block header is initialized by the block proposer (not the runtime). Then we could remove the timestamp pallet altogether which would reduce the coupling.

> This isn't possible without forking the PoW consensus crate, there's zero possible configuration.

This would probably be a good idea in the long run anyways. But I’ll leave it up to you whether we want to go with this approach.

**Comment by CodeSandwich** [comment]
We would need to fork the PoW, then we could drop the timestamp pallet with its inherent in favor of an in-house creation and verification mechanism of a digest item. That would definitely be cleaner, but I don't think that it's worth that kind of effort.

**Pull request merged**